packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -411,11 +427,20 @@ export class ConnectionManager implements IConnectionManager {
     private async connectCore(connectionMode?: ConnectionMode): Promise<void> {
         assert(!this.closed, 0x26a /* "not closed" */);
 
-        if (this.connection !== undefined || this.pendingConnection) {
-            return;
+        if (this.connection !== undefined) {
+            return;  // Connection attempt already completed successfully
+        }
+
+        let pendingConnectionMode;
+        if (this.pendingConnection !== undefined) {
+            pendingConnectionMode = this.pendingConnection.connectionMode;
+            this.cancelConnection();  // Throw out in-progress connection attempt in favor of new attempt
         }
```

Comment:
Maybe assert pendingConnection is now undefined just to be super clear (or perhaps a comment would be sufficient)